### PR TITLE
Deprecate `collection.mutable.AnyRefMap`

### PR DESCRIPTION
**Pull request by wangyum**
Since 2.13, `mutable.AnyRefMap` no faster than `mutable.HashMap`. Please see the benchmark: https://github.com/scala/bug/issues/13029#issuecomment-2325489902. This PR deprecates `mutable.AnyRefMap`.

### CODE REVIEW

**Comment by som-snytt** [comment]
The companion must be deprecated (since https://github.com/scala/scala/pull/10311).

The implicit `anyRefIsMap` can be deprecated, too.

**Comment by SethTisue** [comment]
@wangyum interested in returning to this...?

@scala/collections

**Comment by SethTisue** [comment]
I added some additional `@nowarn`s (because `fatalWarnings` is false locally, true on CI)

**Review by SethTisue — APPROVED** [review]
(assuming CI likes it)

File: src/library/scala/collection/mutable/AnyRefMap.scala [hunk 1]
```diff
@@ -41,6 +42,7 @@ import scala.language.implicitConversions
  *  rapidly as 2^30^ is approached.
  *
  */
+@(deprecated @companionClass)("Use `scala.collection.mutable.HashMap` instead for better performance.", since = "2.13.0")
 class AnyRefMap[K <: AnyRef, V] private[collection] (defaultEntry: K => V, initialBufferSize: Int, initBlank: Boolean)
```
Comment (He-Pin) [inline src/library/scala/collection/mutable/AnyRefMap.scala:46]: since 2.13.16 ?

File: src/library/scala/collection/mutable/AnyRefMap.scala [hunk 2]
```diff
@@ -41,6 +42,7 @@ import scala.language.implicitConversions
  *  rapidly as 2^30^ is approached.
  *
  */
+@(deprecated @companionClass)("Use `scala.collection.mutable.HashMap` instead for better performance.", since = "2.13.0")
```
Comment (He-Pin) [inline src/library/scala/collection/mutable/AnyRefMap.scala]: since 2.13.16 here?

File: src/library/scala/collection/mutable/AnyRefMap.scala [hunk 3]
```diff
@@ -521,6 +523,7 @@ class AnyRefMap[K <: AnyRef, V] private[collection] (defaultEntry: K => V, initi
   override protected[this] def stringPrefix = "AnyRefMap"
 }
 
+@deprecated("Use `scala.collection.mutable.HashMap` instead for better performance.", since = "2.13.0")
```
Comment (He-Pin) [inline src/library/scala/collection/mutable/AnyRefMap.scala]: 2.13.16

**Comment by SethTisue** [comment]
@He-Pin nice catch! fixed.

**Comment by SethTisue** [comment]
Thank you @wangyum!

**Comment by bishabosha** [comment]
I don't think this should be deprecated as it does have significant savings on memory usage

**Comment by wangyum** [comment]
@bishabosha `HashMap` is much master than `AnyRefMap` since Scala 2.13.

**Comment by Ichoran** [comment]
@wangyum - But Jamie's point was about _memory usage_, not speed.  Sometimes memory is what is limiting.

The question is then: how much of a difference, and is it worth maintaining a separate implementation in order to support a lower-memory variant?

**Comment by bishabosha** [comment]
AnyRefMap does not allocate any object to store a key/value pair - so as long as you stay within a size bound then it amortizes to just the initial array allocation. Great for pooling if you clear the map frequently

**Pull request merged**